pkg/workload/tpcc/new_order.go

```diff
@@ -205,232 +207,264 @@ func (n *newOrder) run(ctx context.Context, wID int) (interface{}, time.Duration
-					strings.Join(sOrderCntUpdateCases, " "),
-					strings.Join(sRemoteCntUpdateCases, " "),
-					strings.Join(stockIDs, ", "),
-				),
-			); err != nil {
-				return errors.Wrap(err, "update stock failed")
-			}
+						strings.Join(sQuantityUpdateCases, " "),
+						strings.Join(sYtdUpdateCases, " "),
+						strings.Join(sOrderCntUpdateCases, " "),
+						strings.Join(sRemoteCntUpdateCases, " "),
+						strings.Join(stockIDs, ", "),
+					),
+				); err != nil {
+					return errors.Wrap(err, "update stock failed")
+				}
 
-			// Insert a new order line for each item in the order.
-			olValsStrings := make([]string, d.oOlCnt)
-			for i := range d.items {
-				item := &d.items[i]
-				item.olAmount = float64(item.olQuantity) * item.iPrice
-				d.totalAmount += item.olAmount
-
-				olValsStrings[i] = fmt.Sprintf("(%d,%d,%d,%d,%d,%d,%d,%f,'%s')",
-					d.oID,            // ol_o_id
-					d.dID,            // ol_d_id
-					d.wID,            // ol_w_id
-					item.olNumber,    // ol_number
-					item.olIID,       // ol_i_id
-					item.olSupplyWID, // ol_supply_w_id
-					item.olQuantity,  // ol_quantity
-					item.olAmount,    // ol_amount
-					distInfos[i],     // ol_dist_info
-				)
-			}
-			if _, err := tx.Exec(
-				ctx,
-				fmt.Sprintf(`
+				// Insert a new order line for each item in the order.
+				olValsStrings := make([]string, d.oOlCnt)
+				for i := range d.items {
+					item := &d.items[i]
+					item.olAmount = float64(item.olQuantity) * item.iPrice
+					d.totalAmount += item.olAmount
+
+					olValsStrings[i] = fmt.Sprintf("(%d,%d,%d,%d,%d,%d,%d,%f,'%s')",
+						d.oID,            // ol_o_id
+						d.dID,            // ol_d_id
+						d.wID,            // ol_w_id
+						item.olNumber,    // ol_number
+						item.olIID,       // ol_i_id
+						item.olSupplyWID, // ol_supply_w_id
+						item.olQuantity,  // ol_quantity
+						item.olAmount,    // ol_amount
+						distInfos[i],     // ol_dist_info
+					)
+				}
+				if _, err := tx.Exec(
+					ctx,
+					fmt.Sprintf(`
 					INSERT INTO order_line(ol_o_id, ol_d_id, ol_w_id, ol_number, ol_i_id, ol_supply_w_id, ol_quantity, ol_amount, ol_dist_info)
 					VALUES %s`,
-					strings.Join(olValsStrings, ", "),
-				),
-			); err != nil {
-				return errors.Wrap(err, "insert order_line failed")
-			}
+						strings.Join(olValsStrings, ", "),
+					),
+				); err != nil {
+					return errors.Wrap(err, "insert order_line failed")
+				}
 
-			// 2.4.2.2: total_amount = sum(OL_AMOUNT) * (1 - C_DISCOUNT) * (1 + W_TAX + D_TAX)
-			d.totalAmount *= (1 - d.cDiscount) * (1 + d.wTax + d.dTax)
+				// 2.4.2.2: total_amount = sum(OL_AMOUNT) * (1 - C_DISCOUNT) * (1 + W_TAX + D_TAX)
+				d.totalAmount *= (1 - d.cDiscount) * (1 + d.wTax + d.dTax)
+
+				return nil
+			})
+		if errors.Is(err, errSimulated) {
+			return d, 0, nil
+		}
+
+		if err != nil && n.config.repairOrderIds {
+			var pgErr *pgconn.PgError
+			isUniqErr := errors.As(err, &pgErr) && pgcode.MakeCode(pgErr.Code) == pgcode.UniqueViolation
```

Comment:
i don't think i understand the full context of the problem you're trying to solve. Why does LDR lead to a unique violation in the foreground workload here? are we getting a unique violation on a the district pk? how exactly would this retry behavior help LDR performance?

Reply:
`new_order` txn inserting into order is what fails, because the order id picked already exists, because it is picked by something that looks like a sequence, specifically the  order id is picked by incrementing the “next_order_id” column for the corresponding district. 

Suppose both clusters have a next_id of 5 and our replication latency is 10.

At times 1 and 2, Cluster A processes 2 orders, 5 and 6, incrementing it to 7. 

At time 4, cluster B — which has not yet replicated any of A’s writes, processes one order, also giving it ID 5 and setting its next id — at time 4 — to 6.

When B hears about A’s update of next to be 6, it is newer, since it happened at t4, than its local update to 7 at t2, so it wins LWW and next regresses to 6. Meanwhile it replaces the order with ID 6 since why not — no conflict there.

now both clusters have a next ID of 6, but have an order with ID 6, so they cannot accept any future orders for that warehouse and district until someone “repairs” the next order id field, as we now do here.

Reply:
i see, thanks for the explanation. i think this blurb should probably live in the code or in the commit message.